x/posts/types/models.go

```diff
@@ -0,0 +1,669 @@
+	}
+
+	return nil
+}
+
+// --------------------------------------------------------------------------------------------------------------------
+
+// PostUpdate contains all the data that can be updated about a post.
+// When performing an update, if a text field should not be edited then it must be set to types.DoNotModify.
+type PostUpdate struct {
+	// If it shouldn't replace the current text, it must be set to types.DoNotModify
+	Text string
+
+	// Update's entities will always replace the existing ones
+	Entities *Entities
+
+	UpdateTime time.Time
+}
+
+// NewPostUpdate returns a new PostUpdate instance
+func NewPostUpdate(text string, entities *Entities, updateTime time.Time) PostUpdate {
+	return PostUpdate{
+		Text:       text,
+		Entities:   entities,
+		UpdateTime: updateTime,
+	}
+}
+
+// Update updates the fields of a given post without validating it.
+// Before storing the updated post, a validation with keeper.ValidatePost should
+// be performed.
+func (p Post) Update(update PostUpdate) Post {
+	if update.Text == DoNotModify {
+		update.Text = p.Text
+	}
+
+	return NewPost(
+		p.SubspaceID,
+		p.ID,
+		p.ExternalID,
+		update.Text,
+		p.Author,
+		p.ConversationID,
+		update.Entities,
+		p.ReferencedPosts,
+		p.ReplySettings,
+		p.CreationDate,
+		&update.UpdateTime,
+	)
+}
+
+// --------------------------------------------------------------------------------------------------------------------
+
+// NewEntities returns a new Entities instance
+func NewEntities(hashtags []Tag, mentions []Tag, urls []Url) *Entities {
+	return &Entities{
+		Hashtags: hashtags,
+		Mentions: mentions,
+		Urls:     urls,
+	}
+}
+
+// Validate implements fmt.Validator
+func (e *Entities) Validate() error {
+	for _, tag := range e.Hashtags {
+		err := tag.Validate()
+		if err != nil {
+			return fmt.Errorf("invalid hashtag: %s", err)
+		}
+	}
+
+	for _, tag := range e.Mentions {
+		err := tag.Validate()
+		if err != nil {
+			return fmt.Errorf("invalid mention: %s", err)
+		}
+	}
+
+	for _, url := range e.Urls {
+		err := url.Validate()
+		if err != nil {
+			return fmt.Errorf("invalid url: %s", err)
+		}
+	}
+
+	// --- Make sure there are no overlapping entities based on (start, end) ---
+
+	// Map all entities into segments
+	segments := e.getSegments()
+
+	if len(segments) < 1 {
+		// We cannot have an empty entities here
+		return fmt.Errorf("entities must have at least one entity inside")
+	}
+
+	if len(segments) < 2 {
+		// With less than 2 segments there cannot be any overlap
+		return nil
+	}
+
+	// Sort the segments
+	sort.Slice(segments, func(i, j int) bool {
+		return segments[i].start < segments[j].start
+	})
+
+	// Verify there are no overlapping segments
+	for index := 0; index < len(segments)-1; index++ {
+		first, second := segments[index], segments[index+1]
+		if first.end >= second.start {
+			return fmt.Errorf("entities cannot overlap: start %d end %d", first.end, second.start)
+		}
+	}
+
+	return nil
+}
+
+type entitySegment struct {
```

Comment:
Can we add docs to this struct and generate it with protobuf instead of manually coding it?

Reply:
I've added the documentation, but I don't think this should be generated using Protobuf. From the documentation, you will see that it's only a type definition that should be private within the `types` package as it's only used temporarily during the validation of an `Entity` object. It's not used elsewhere, hence the decision of making it private and not accessible from anywhere else